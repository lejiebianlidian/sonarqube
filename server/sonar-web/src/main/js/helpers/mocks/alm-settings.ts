/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import {
  AlmKeys,
  AlmSettingsInstance,
  AzureBindingDefinition,
  BitbucketBindingDefinition,
  GithubBindingDefinition,
  GitlabBindingDefinition,
  ProjectAlmBindingResponse,
  ProjectBitbucketBindingResponse
} from '../../types/alm-settings';

export function mockAlmSettingsInstance(
  overrides: Partial<AlmSettingsInstance> = {}
): AlmSettingsInstance {
  return {
    alm: AlmKeys.GitHub,
    key: 'key',
    ...overrides
  };
}

export function mockAzureBindingDefinition(
  overrides: Partial<AzureBindingDefinition> = {}
): AzureBindingDefinition {
  return {
    key: 'key',
    personalAccessToken: 'asdf1234',
    ...overrides
  };
}

export function mockBitbucketBindingDefinition(
  overrides: Partial<BitbucketBindingDefinition> = {}
): BitbucketBindingDefinition {
  return {
    key: 'key',
    personalAccessToken: 'asdf1234',
    url: 'http://bbs.enterprise.com',
    ...overrides
  };
}

export function mockGithubBindingDefinition(
  overrides: Partial<GithubBindingDefinition> = {}
): GithubBindingDefinition {
  return {
    key: 'key',
    url: 'http://github.enterprise.com',
    appId: '123456',
    clientId: 'client1',
    clientSecret: '**clientsecret**',
    privateKey: 'asdf1234',
    ...overrides
  };
}

export function mockGitlabBindingDefinition(
  overrides: Partial<GitlabBindingDefinition> = {}
): GitlabBindingDefinition {
  return {
    key: 'foo',
    personalAccessToken: 'foobar',
    ...overrides
  };
}

export function mockProjectAlmBindingResponse(
  overrides: Partial<ProjectAlmBindingResponse> = {}
): ProjectAlmBindingResponse {
  return {
    alm: AlmKeys.GitHub,
    key: 'foo',
    ...overrides
  };
}

export function mockProjectBitbucketBindingGet(
  overrides: Partial<ProjectBitbucketBindingResponse> = {}
): ProjectBitbucketBindingResponse {
  return {
    alm: AlmKeys.Bitbucket,
    key: 'foo',
    repository: 'PROJECT_KEY',
    slug: 'repo-slug',
    ...overrides
  };
}
