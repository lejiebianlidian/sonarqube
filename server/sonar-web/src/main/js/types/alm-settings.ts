/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
export const enum AlmKeys {
  Azure = 'azure',
  Bitbucket = 'bitbucket',
  GitHub = 'github',
  GitLab = 'gitlab'
}

export interface AlmBindingDefinition {
  key: string;
}

export interface AzureBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
}

export interface BitbucketBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url: string;
}

export interface GithubBindingDefinition extends AlmBindingDefinition {
  appId: string;
  clientId: string;
  clientSecret: string;
  privateKey: string;
  url: string;
}

export interface GitlabBindingDefinition extends AlmBindingDefinition {
  personalAccessToken: string;
  url?: string;
}

export interface ProjectAlmBindingResponse {
  alm: AlmKeys;
  key: string;
  repository?: string;
  slug?: string;
  summaryCommentEnabled?: boolean;
}

export interface ProjectBitbucketBindingResponse extends ProjectAlmBindingResponse {
  alm: AlmKeys.Bitbucket;
  repository: string;
  slug: string;
}

export interface ProjectAlmBindingParams {
  almSetting: string;
  project: string;
}

export interface AzureProjectAlmBindingParams extends ProjectAlmBindingParams {}

export interface BitbucketProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
  slug: string;
}

export interface GithubProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository: string;
  summaryCommentEnabled: boolean;
}

export interface GitlabProjectAlmBindingParams extends ProjectAlmBindingParams {
  repository?: string;
}

export interface AlmSettingsInstance {
  alm: AlmKeys;
  key: string;
  url?: string;
}

export interface AlmSettingsBindingDefinitions {
  [AlmKeys.Azure]: AzureBindingDefinition[];
  [AlmKeys.Bitbucket]: BitbucketBindingDefinition[];
  [AlmKeys.GitHub]: GithubBindingDefinition[];
  [AlmKeys.GitLab]: GitlabBindingDefinition[];
}
