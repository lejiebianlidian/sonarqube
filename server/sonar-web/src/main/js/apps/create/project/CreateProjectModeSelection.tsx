/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as classNames from 'classnames';
import * as React from 'react';
import HelpTooltip from 'sonar-ui-common/components/controls/HelpTooltip';
import { translate, translateWithParameters } from 'sonar-ui-common/helpers/l10n';
import { getBaseUrl } from 'sonar-ui-common/helpers/urls';
import { AlmKeys } from '../../../types/alm-settings';
import { CreateProjectModes } from './types';

export interface CreateProjectModeSelectionProps {
  almCounts: { [key in AlmKeys]: number };
  loadingBindings: boolean;
  onSelectMode: (mode: CreateProjectModes) => void;
}

function renderAlmOption(
  props: CreateProjectModeSelectionProps,
  alm: AlmKeys,
  mode: CreateProjectModes
) {
  const { almCounts, loadingBindings } = props;

  const count = almCounts[alm];
  const disabled = count !== 1 || loadingBindings;

  return (
    <button
      className={classNames(
        'button button-huge big-spacer-left display-flex-column create-project-mode-type-bbs',
        { disabled }
      )}
      disabled={disabled}
      onClick={() => props.onSelectMode(mode)}
      type="button">
      <img
        alt="" // Should be ignored by screen readers
        height={80}
        src={`${getBaseUrl()}/images/alm/${alm}.svg`}
      />
      <div className="medium big-spacer-top">
        {translate('onboarding.create_project.select_method', alm)}
      </div>

      {loadingBindings && (
        <span>
          {translate('onboarding.create_project.check_alm_supported')}
          <i className="little-spacer-left spinner" />
        </span>
      )}

      {!loadingBindings && disabled && (
        <div className="text-muted small spacer-top" style={{ lineHeight: 1.5 }}>
          {translate('onboarding.create_project.alm_not_configured')}
          <HelpTooltip
            className="little-spacer-left"
            overlay={
              count === 0
                ? translate('onboarding.create_project.zero_alm_instances', alm)
                : `${translate('onboarding.create_project.too_many_alm_instances', alm)} 
                  ${translateWithParameters(
                    'onboarding.create_project.alm_instances_count_X',
                    count
                  )}`
            }
          />
        </div>
      )}
    </button>
  );
}

export default function CreateProjectModeSelection(props: CreateProjectModeSelectionProps) {
  return (
    <>
      <header className="huge-spacer-top big-spacer-bottom padded">
        <h1 className="text-center huge big-spacer-bottom">
          {translate('my_account.create_new.TRK')}
        </h1>
        <p className="text-center big">{translate('onboarding.create_project.select_method')}</p>
      </header>

      <div className="create-project-modes huge-spacer-top display-flex-justify-center">
        <button
          className="button button-huge display-flex-column create-project-mode-type-manual"
          onClick={() => props.onSelectMode(CreateProjectModes.Manual)}
          type="button">
          <img
            alt="" // Should be ignored by screen readers
            height={80}
            src={`${getBaseUrl()}/images/sonarcloud/analysis/manual.svg`}
          />
          <div className="medium big-spacer-top">
            {translate('onboarding.create_project.select_method.manual')}
          </div>
        </button>

        {renderAlmOption(props, AlmKeys.Bitbucket, CreateProjectModes.BitbucketServer)}
        {renderAlmOption(props, AlmKeys.GitHub, CreateProjectModes.GitHub)}
      </div>
    </>
  );
}
