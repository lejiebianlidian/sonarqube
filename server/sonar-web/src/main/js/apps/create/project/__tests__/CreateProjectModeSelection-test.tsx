/*
 * SonarQube
 * Copyright (C) 2009-2020 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { shallow } from 'enzyme';
import * as React from 'react';
import { click } from 'sonar-ui-common/helpers/testUtils';
import { AlmKeys } from '../../../../types/alm-settings';
import CreateProjectModeSelection, {
  CreateProjectModeSelectionProps
} from '../CreateProjectModeSelection';
import { CreateProjectModes } from '../types';

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
  expect(shallowRender({ loadingBindings: true })).toMatchSnapshot('loading instances');
  expect(shallowRender({}, { [AlmKeys.Bitbucket]: 0, [AlmKeys.GitHub]: 2 })).toMatchSnapshot(
    'invalid configs'
  );
});

it('should correctly pass the selected mode up', () => {
  const onSelectMode = jest.fn();
  const wrapper = shallowRender({ onSelectMode });

  click(wrapper.find('button.create-project-mode-type-manual'));
  expect(onSelectMode).toBeCalledWith(CreateProjectModes.Manual);

  click(wrapper.find('button.create-project-mode-type-bbs').at(0));
  expect(onSelectMode).toBeCalledWith(CreateProjectModes.BitbucketServer);

  click(wrapper.find('button.create-project-mode-type-bbs').at(1));
  expect(onSelectMode).toBeCalledWith(CreateProjectModes.GitHub);
});

function shallowRender(
  props: Partial<CreateProjectModeSelectionProps> = {},
  almCountOverrides = {}
) {
  const almCounts = {
    [AlmKeys.Azure]: 0,
    [AlmKeys.Bitbucket]: 1,
    [AlmKeys.GitHub]: 0,
    [AlmKeys.GitLab]: 0,
    ...almCountOverrides
  };
  return shallow<CreateProjectModeSelectionProps>(
    <CreateProjectModeSelection
      almCounts={almCounts}
      loadingBindings={false}
      onSelectMode={jest.fn()}
      {...props}
    />
  );
}
