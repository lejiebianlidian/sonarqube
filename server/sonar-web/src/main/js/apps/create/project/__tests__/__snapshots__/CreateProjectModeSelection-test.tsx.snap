// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: default 1`] = `
<Fragment>
  <header
    className="huge-spacer-top big-spacer-bottom padded"
  >
    <h1
      className="text-center huge big-spacer-bottom"
    >
      my_account.create_new.TRK
    </h1>
    <p
      className="text-center big"
    >
      onboarding.create_project.select_method
    </p>
  </header>
  <div
    className="create-project-modes huge-spacer-top display-flex-justify-center"
  >
    <button
      className="button button-huge display-flex-column create-project-mode-type-manual"
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        height={80}
        src="/images/sonarcloud/analysis/manual.svg"
      />
      <div
        className="medium big-spacer-top"
      >
        onboarding.create_project.select_method.manual
      </div>
    </button>
    <button
      className="button button-huge big-spacer-left display-flex-column create-project-mode-type-bbs"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        height={80}
        src="/images/alm/bitbucket.svg"
      />
      <div
        className="medium big-spacer-top"
      >
        onboarding.create_project.select_method.bitbucket
      </div>
    </button>
    <button
      className="button button-huge big-spacer-left display-flex-column create-project-mode-type-bbs disabled"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        height={80}
        src="/images/alm/github.svg"
      />
      <div
        className="medium big-spacer-top"
      >
        onboarding.create_project.select_method.github
      </div>
      <div
        className="text-muted small spacer-top"
        style={
          Object {
            "lineHeight": 1.5,
          }
        }
      >
        onboarding.create_project.alm_not_configured
        <HelpTooltip
          className="little-spacer-left"
          overlay="onboarding.create_project.zero_alm_instances.github"
        />
      </div>
    </button>
  </div>
</Fragment>
`;

exports[`should render correctly: invalid configs 1`] = `
<Fragment>
  <header
    className="huge-spacer-top big-spacer-bottom padded"
  >
    <h1
      className="text-center huge big-spacer-bottom"
    >
      my_account.create_new.TRK
    </h1>
    <p
      className="text-center big"
    >
      onboarding.create_project.select_method
    </p>
  </header>
  <div
    className="create-project-modes huge-spacer-top display-flex-justify-center"
  >
    <button
      className="button button-huge display-flex-column create-project-mode-type-manual"
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        height={80}
        src="/images/sonarcloud/analysis/manual.svg"
      />
      <div
        className="medium big-spacer-top"
      >
        onboarding.create_project.select_method.manual
      </div>
    </button>
    <button
      className="button button-huge big-spacer-left display-flex-column create-project-mode-type-bbs disabled"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        height={80}
        src="/images/alm/bitbucket.svg"
      />
      <div
        className="medium big-spacer-top"
      >
        onboarding.create_project.select_method.bitbucket
      </div>
      <div
        className="text-muted small spacer-top"
        style={
          Object {
            "lineHeight": 1.5,
          }
        }
      >
        onboarding.create_project.alm_not_configured
        <HelpTooltip
          className="little-spacer-left"
          overlay="onboarding.create_project.zero_alm_instances.bitbucket"
        />
      </div>
    </button>
    <button
      className="button button-huge big-spacer-left display-flex-column create-project-mode-type-bbs disabled"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        height={80}
        src="/images/alm/github.svg"
      />
      <div
        className="medium big-spacer-top"
      >
        onboarding.create_project.select_method.github
      </div>
      <div
        className="text-muted small spacer-top"
        style={
          Object {
            "lineHeight": 1.5,
          }
        }
      >
        onboarding.create_project.alm_not_configured
        <HelpTooltip
          className="little-spacer-left"
          overlay="onboarding.create_project.too_many_alm_instances.github 
                  onboarding.create_project.alm_instances_count_X.2"
        />
      </div>
    </button>
  </div>
</Fragment>
`;

exports[`should render correctly: loading instances 1`] = `
<Fragment>
  <header
    className="huge-spacer-top big-spacer-bottom padded"
  >
    <h1
      className="text-center huge big-spacer-bottom"
    >
      my_account.create_new.TRK
    </h1>
    <p
      className="text-center big"
    >
      onboarding.create_project.select_method
    </p>
  </header>
  <div
    className="create-project-modes huge-spacer-top display-flex-justify-center"
  >
    <button
      className="button button-huge display-flex-column create-project-mode-type-manual"
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        height={80}
        src="/images/sonarcloud/analysis/manual.svg"
      />
      <div
        className="medium big-spacer-top"
      >
        onboarding.create_project.select_method.manual
      </div>
    </button>
    <button
      className="button button-huge big-spacer-left display-flex-column create-project-mode-type-bbs disabled"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        height={80}
        src="/images/alm/bitbucket.svg"
      />
      <div
        className="medium big-spacer-top"
      >
        onboarding.create_project.select_method.bitbucket
      </div>
      <span>
        onboarding.create_project.check_alm_supported
        <i
          className="little-spacer-left spinner"
        />
      </span>
    </button>
    <button
      className="button button-huge big-spacer-left display-flex-column create-project-mode-type-bbs disabled"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <img
        alt=""
        height={80}
        src="/images/alm/github.svg"
      />
      <div
        className="medium big-spacer-top"
      >
        onboarding.create_project.select_method.github
      </div>
      <span>
        onboarding.create_project.check_alm_supported
        <i
          className="little-spacer-left spinner"
        />
      </span>
    </button>
  </div>
</Fragment>
`;
